torchvision/prototype/models/depth/stereo/raft_stereo.py

```diff
@@ -0,0 +1,746 @@
+                nn.init.kaiming_normal_(m.weight, mode="fan_out", nonlinearity="relu")
+            elif isinstance(m, (nn.BatchNorm2d, nn.InstanceNorm2d)):
+                if m.weight is not None:
+                    nn.init.constant_(m.weight, 1)
+                if m.bias is not None:
+                    nn.init.constant_(m.bias, 0)
+
+        num_downsampling = sum([x - 1 for x in strides])
+        self.downsampling_ratio = 2 ** (num_downsampling)
+
+    def _make_2_blocks(self, block, in_channels, out_channels, norm_layer, first_stride):
+        block1 = block(in_channels, out_channels, norm_layer=norm_layer, stride=first_stride)
+        block2 = block(out_channels, out_channels, norm_layer=norm_layer, stride=1)
+        return nn.Sequential(block1, block2)
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = self.convnormrelu(x)
+
+        x = self.layer1(x)
+        x = self.layer2(x)
+        x = self.layer3(x)
+        return x
+
+
+class FeatureEncoder(nn.Module):
+    def __init__(
+        self,
+        base_encoder: BaseEncoder,
+        output_dim: int = 256,
+        shared_base: bool = False,
+        block: Callable[..., nn.Module] = ResidualBlock,
+    ):
+        super().__init__()
+        self.base_encoder = base_encoder
+        self.base_downsampling_ratio = base_encoder.downsampling_ratio
+        base_dim = base_encoder.output_dim
+
+        # If we share base encoder weight for Feature and Context Encoder
+        # we need to add residual block with InstanceNorm2d and change the kernel size for conv layer
+        # see: https://github.com/princeton-vl/RAFT-Stereo/blob/main/core/raft_stereo.py#L35-L37
+        self.shared_base = shared_base
+        if shared_base:
+            self.residual_block = block(base_dim, base_dim, norm_layer=nn.InstanceNorm2d, stride=1)
+            self.conv = nn.Conv2d(base_dim, output_dim, kernel_size=3, padding=1)
+        else:
+            self.conv = nn.Conv2d(base_dim, output_dim, kernel_size=1)
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = self.base_encoder(x)
+        if self.shared_base:
+            x = self.residual_block(x)
+        x = self.conv(x)
+        return x
+
+
+class MultiLevelContextEncoder(nn.Module):
+    def __init__(
+        self,
+        base_encoder: nn.Module,
+        output_dim: int = 256,
+        out_with_blocks: List[int] = [1, 1, 0],
+        block: Callable[..., nn.Module] = ResidualBlock,
+    ):
+        super().__init__()
+        self.num_level = len(out_with_blocks)
+        self.base_encoder = base_encoder
+        self.base_downsampling_ratio = base_encoder.downsampling_ratio
+        base_dim = base_encoder.output_dim
+
+        # Layer to output hidden_state and context separately (each produce output_dim//2 dims)
+        self.out_hidden_states = nn.ModuleList(
+            [
+                self._make_out_layer(base_dim, output_dim // 2, with_block=with_block, block=block)
```

Comment:
Optional Nit: We rarely do the construction of the layers in private methods any more as it hurts readability. Consider simplifying the logic of the method by always building sequential layers and moving the construction in the constructor.

Reply:
Thanks, will try to simplify this.